File: Week_09/85/maximal_rectangle.py
from typing import List


class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if not matrix or not matrix[0]:
            return 0
        m, n = len(matrix), len(matrix[0])

        height = [0] * n
        left = [-1] * n
        right = [n] * n
        res = 0
        for i in range(m):
            cur_left, cur_right = -1, n
            for j in range(n):
                if matrix[i][j] == '1':
                    height[j] += 1
                else:
                    height[j] = 0

            for j in range(n):
                if matrix[i][j] == '1':
                    left[j] = max(left[j], cur_left)
                else:
                    left[j] = -1
                    cur_left = j
            for j in range(n-1, -1, -1):
                if matrix[i][j] == '1':
                    right[j] = min(right[j], cur_right)
                else:
                    right[j] = n
                    cur_right = j
            for j in range(n):
                res = max(res, height[j] * (right[j] - left[j] - 1))
        return res

File: Week_09/85/maximal_rectangle.go
package main

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func maximalRectangle(matrix [][]byte) int {
	m := len(matrix)
	if m == 0 {
		return 0
	}
	n := len(matrix[0])
	if n == 0 {
		return 0
	}

	left := make([]int, n)
	for j := 0; j < n; j++ {
		left[j] = -1
	}
	right := make([]int, n)
	for j := 0; j < n; j++ {
		right[j] = n
	}
	height := make([]int, n)

	res := 0

	for i := 0; i < m; i++ {
		curLeft, curRight := -1, n
		//height
		for j := 0; j < n; j++ {
			if matrix[i][j] == '1' {
				height[j]++
			} else {
				height[j] = 0
			}
		}
		//left
		for j := 0; j < n; j++ {
			if matrix[i][j] == '1' {
				left[j] = max(left[j], curLeft)
			} else {
				left[j] = -1
				curLeft = j
			}
		}
		//right
		for j := n - 1; j >= 0; j-- {
			if matrix[i][j] == '1' {
				right[j] = min(right[j], curRight)
			} else {
				right[j] = n
				curRight = j
			}
		}
		//area
		for j := 0; j < n; j++ {
			res = max(res, height[j]*(right[j]-left[j]-1))
		}
		//fmt.Printf("%d:  \n%v\n%v\n", i, left, right)
		//for j := 0; j < n; j++ {
		//	fmt.Printf("j: %d, left: %d, right: %d, area: %d\n", j, left[j], right[j], height[j]*(right[j]-left[j]-1))
		//}
	}
	return res
}
